### ISSUE TITLE
A better parser for logfiles

### ISSUE THREAD

**Issue by Freed-Wu** [vim/vim#11434]
**Is your feature request about something that is currently impossible or hard 
to do? Please describe the problem.**

<https://github.com/lervag/vimtex/issues/2526#issuecomment-1288686446>

Since now, vim's 'errorformat' is hard to parse this situation:

```
1 ./a.tex:15: Undefined control sequence.
2 \beamer@doifinframe ...bular}{lcccccc} \cdashline
3                                                   {1-7} \end {tabular} \end ...
4 l.15 \end{frame}
```

So can we have a better errorformat to handle this situation?

**Comment by bfrg** [comment]
What does your current `errorformat` look like? And which of the lines 2-4 do you need to add to the first one? Does the 4th line signal the end of the multi-line error message, or is the `l.15` a coincidence?

**Comment by Freed-Wu** [comment]
@lervag <https://github.com/lervag/vimtex/tree/master/compiler> seems that don't have latexmk's compiler?

**Comment by lervag** [comment]
I believe the main point here is that the `errorformat` "language" is restricted in the sense that there is no state between lines. That is, when we parse more or less unstructured logs such as those from LaTeX, `errorformat` is not always sufficient because the content of line 2 in the example above can be more or less random.

VimTeX defines the errorformat here:

https://github.com/lervag/vimtex/blob/4e92ce6/autoload/vimtex/qf/latexlog.vim#L25

Line 1 in the example is matched by

https://github.com/lervag/vimtex/blob/4e92ce6/autoload/vimtex/qf/latexlog.vim#L37

However, there is no way to do a continuation for a "random" line unless we implement it as _any_ continuation. That is, the `%C` will match a continuation to any _active_ multiline error/warning/info.

**Comment by bfrg** [comment]
I don't think there's a way to accomplish this without something like `%C%.%#` either.

I haven't worked with Latex in over 8 years, so I don't remember its output format. But if multi-line error messages end with `l.15`, where the 15 is the line number where the error occurs (i.e. it signals the end of the multi-line error message), then the following could be used:
```vim
vim9script

const text: list<string> =<< trim END
    ./a.tex:15: Undefined control sequence.
    \beamer@doifinframe ...bular}{lcccccc} \cdashline
                                                    {1-7} \end {tabular} \end ...
    l.15 \end{frame}
END

const efm: list<string> =<< trim END
    %E%f:%l: %m
    %+Zl.%*\d%.%#
    %+C%.%#
END

setqflist([], ' ', {lines: text, efm: efm->join(',')->escape(' \')})
```

You have to make sure the `%C%.%#` is at the very end of the `errorformat` so that it only matches when everything else fails.

I use something like this for my `clang` errorformat and so far it works fine.

**Comment by lervag** [comment]
> I don't think there's a way to accomplish this without something like `%C%.%#` either.

Yes, we can accomplish it with `%C%.%#`, but that will have the side effect that it will add noise to many other errors/warnings. That's why we can't actually do that.

> You have to make sure the `%C%.%#` is at the very end of the `errorformat` so that it only matches when everything else fails.

Yes, but for the same reason as explained above, we can't do this. To be more specific, the problem is that we have to ignore lines that don't match (with something like `%-G%.%#`), because many multi-line messages do not have a decisive _end_ line that we can match with `%Z`.

**Comment by brammool** [comment]
> I don't think there's a way to accomplish this without something like `%C%.%#` either.

The 'errorformat' option was intended for simple cases.
More features have been added for more complicated things, but it makes
the value hard to use.  You may very well prefer to parse the text in
some other way and use setqflist() to store the result.
That way you can do anything and write the parser in a clean way.


-- 
For society, it's probably a good thing that engineers value function over
appearance.  For example, you wouldn't want engineers to build nuclear power
plants that only _look_ like they would keep all the radiation inside.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lervag** [comment]
> You may very well prefer to parse the text in some other way and use setqflist() to store the result. That way you can do anything and write the parser in a clean way.

Yes, I think this is a sensible route and a good suggestion. In fact, I think it may answer the original question in the sense that 1) expanding the current `'errorformat'` seems way too complicated for the present issue, and 2) writing a custom parser and using `setqflist()` allows more or less full flexibility.

Now, writing such a parser for LaTeX log files is still hard and I will not take that task on myself. Sorry :)

**Comment by chrisbra** [comment]
I think writing a proper LaTeX log file parser is not really suitable for the Vim project. So closing the ticket

**Comment by lervag** [comment]
> I think writing a proper LaTeX log file parser is not really suitable for the Vim project. So closing the ticket

I would claim that 1) the suggested `errorformat` in `:help errorformat-LaTeX` is already quite good, and 2) the version provided by VimTeX is even better. It is not perfect, then parsing the LaTeX log file format is not trivial. We could do better, and it would be possible to do that both in Vimscript or any external tooling and combine it with `setqflist()`. This should be left to the community. I don't have the time, but if anyone works on this, then I would not mind to consider a PR for VimTeX.

Or, in short: I agree with closing the ticket.
